File: src/lib/media/store.ts
/**
 * Media store — badhi image / video / audio file ahiya thi j pasar thay che.
 *
 * Be jagya e rahe che:
 *   disk   — `storage/media/...` — render pipeline ne local file joiye che
 *   public — Cloudinary/Catbox/etc — Meta ne download karva mate URL joiye che
 *
 * Etle "save karo" ek j call che, ane pachi jarur pade tyare public URL
 * lazily banave che.
 */

import { randomUUID } from "node:crypto";
import { mkdir, readFile, writeFile, unlink, stat } from "node:fs/promises";
import { existsSync } from "node:fs";
import path from "node:path";
import sharp from "sharp";

import { connectDB } from "@/lib/db";
import {
  MediaAsset,
  type MediaAssetDoc,
  type MediaAssetDocument,
} from "@/models/MediaAsset";
import { uploadPublic } from "./hosts";
import { apiFetch } from "@/lib/pipeline/chain";

export type MediaKind = "image" | "video" | "audio" | "other";
export type MediaRole = MediaAssetDoc["role"];

/** Badhi file ahiya rahe che — .gitignore ma che. */
export function mediaRoot(): string {
  return process.env.MEDIA_DIR
    ? path.resolve(process.env.MEDIA_DIR)
    : path.join(process.cwd(), "storage", "media");
}

/** Render vakhate vaparata temporary file mate. */
export function workRoot(): string {
  return path.join(mediaRoot(), "..", "work");
}

export async function ensureDir(dir: string): Promise<void> {
  if (!existsSync(dir)) await mkdir(dir, { recursive: true });
}

const EXTENSIONS: Record<string, string> = {
  "image/jpeg": ".jpg",
  "image/jpg": ".jpg",
  "image/png": ".png",
  "image/webp": ".webp",
  "image/gif": ".gif",
  "video/mp4": ".mp4",
  "video/quicktime": ".mov",
  "video/webm": ".webm",
  "audio/mpeg": ".mp3",
  "audio/mp4": ".m4a",
  "audio/wav": ".wav",
  "audio/ogg": ".ogg",
};

export function extensionFor(mimeType: string, fallback = ".bin"): string {
  return EXTENSIONS[mimeType.toLowerCase().split(";")[0].trim()] ?? fallback;
}

export function kindFor(mimeType: string): MediaKind {
  const type = mimeType.toLowerCase();
  if (type.startsWith("image/")) return "image";
  if (type.startsWith("video/")) return "video";
  if (type.startsWith("audio/")) return "audio";
  return "other";
}

/** Filename ne surakshit banave — path traversal band. */
function safeName(name: string): string {
  return (
    path
      .basename(name)
      .replace(/[^a-zA-Z0-9._-]/g, "_")
      .slice(0, 80) || "file"
  );
}

export type SaveMediaInput = {
  data: Buffer;
  filename?: string;
  mimeType: string;
  role?: MediaRole;
  brand?: string;
  createdBy?: string;
  provider?: string;
  prompt?: string;
  duration?: number;
  width?: number;
  height?: number;
  /**
   * true  = turant public host par pan chadhavo (Meta ne aapva mate)
   * false = fakt disk par (intermediate file mate)
   */
  makePublic?: boolean;
  preferHost?: string;
};

/**
 * Buffer ne disk par lakhe, DB ma record banave, ane (kahyu hoy to)
 * public URL pan kadhi aape.
 */
export async function saveMedia(input: SaveMediaInput): Promise<MediaAssetDocument> {
  await connectDB();

  const kind = kindFor(input.mimeType);
  const ext = extensionFor(input.mimeType, path.extname(input.filename ?? "") || ".bin");
  const id = randomUUID();
  const base = safeName(input.filename ?? `${input.role ?? kind}${ext}`);
  const stored = `${id}-${base.endsWith(ext) ? base : base + ext}`;

  const bucket = new Date().toISOString().slice(0, 7); // 2026-08
  const dir = path.join(mediaRoot(), bucket);
  await ensureDir(dir);

  const localPath = path.join(dir, stored);
  await writeFile(localPath, input.data);

  let { width, height } = input;
  if (kind === "image" && (!width || !height)) {
    try {
      const meta = await sharp(input.data).metadata();
      width = meta.width;
      height = meta.height;
    } catch {
      // Metadata na madyu to vandho nahi — file to save thai gai che.
    }
  }

  const asset = await MediaAsset.create({
    brand: input.brand,
    kind,
    role: input.role ?? "other",
    filename: stored,
    mimeType: input.mimeType,
    bytes: input.data.length,
    localPath,
    width,
    height,
    duration: input.duration,
    provider: input.provider,
    prompt: input.prompt,
    host: "local",
    createdBy: input.createdBy,
  });

  if (input.makePublic) {
    await ensurePublicUrl(asset, input.preferHost);
  }

  return asset;
}

/**
 * Asset ne public URL aapo — nahi hoy to atyare j banavo.
 *
 * Kram:
 *   1. Pehle thi URL che ane hju valid che → e j
 *   2. PUBLIC_MEDIA_BASE_URL set che → aapno potano /api/media route
 *   3. Nahi to external host chain (Cloudinary → ImgBB → Catbox → tmpfiles)
 */
export async function ensurePublicUrl(
  asset: MediaAssetDocument,
  preferHost?: string,
): Promise<string> {
  const stillValid =
    asset.publicUrl &&
    (!asset.expiresAt || asset.expiresAt.getTime() > Date.now() + 60_000);
  if (stillValid) return asset.publicUrl as string;

  const base = (process.env.PUBLIC_MEDIA_BASE_URL || "").replace(/\/$/, "");
  if (base && /^https:\/\//i.test(base)) {
    const url = `${base}/api/media/${asset._id}`;
    asset.publicUrl = url;
    asset.host = "base-url";
    asset.expiresAt = undefined;
    await asset.save();
    return url;
  }

  const data = await readMediaBuffer(asset);
  const uploaded = await uploadPublic(
    {
      data,
      filename: asset.filename,
      mimeType: asset.mimeType,
      kind: asset.kind as MediaKind,
    },
    preferHost,
  );

  asset.publicUrl = uploaded.data.url;
  asset.host = uploaded.data.host;
  asset.expiresAt = uploaded.data.expiresAt;
  await asset.save();

  return uploaded.data.url;
}

/** Disk par thi asset ni file vanche. */
export async function readMediaBuffer(asset: MediaAssetDocument): Promise<Buffer> {
  if (asset.localPath && existsSync(asset.localPath)) {
    return readFile(asset.localPath);
  }
  if (asset.publicUrl) {
    // Local file gum thai gai pan public URL che — tya thi pachi lai laiye.
    return apiFetch<Buffer>(asset.publicUrl, { expect: "buffer" });
  }
  throw new Error(`Media file madi nahi: ${asset.filename}`);
}

/** Asset nu local path aapo — nahi hoy to public URL par thi download karo. */
export async function ensureLocalPath(asset: MediaAssetDocument): Promise<string> {
  if (asset.localPath && existsSync(asset.localPath)) return asset.localPath;

  const data = await readMediaBuffer(asset);
  const dir = path.join(mediaRoot(), "restored");
  await ensureDir(dir);
  const target = path.join(dir, asset.filename);
  await writeFile(target, data);

  asset.localPath = target;
  await asset.save();
  return target;
}

const MAX_DOWNLOAD_BYTES = 200 * 1024 * 1024; // 200MB

/**
 * Bahar na URL par thi file lai ne aapna store ma mukhe.
 * Reference reel ke product page ni image mate vaparay che.
 */
export async function saveMediaFromUrl(
  url: string,
  input: Omit<SaveMediaInput, "data" | "mimeType"> & {
    mimeType?: string;
    /** Khaas headers joita hoy to (hotlink protection valі site mate). */
    headers?: Record<string, string>;
  },
): Promise<MediaAssetDocument> {
  if (!/^https?:\/\//i.test(url)) {
    throw new Error(`Khotu URL: ${url.slice(0, 120)}`);
  }

  // Ghani site hotlink block kare che — Referer vagar 403 aape che.
  // Etle file je site par thi aave che e j site ne Referer tarike aapiye chie.
  const origin = (() => {
    try {
      return new URL(url).origin + "/";
    } catch {
      return undefined;
    }
  })();

  const response = await fetch(url, {
    signal: AbortSignal.timeout(180_000),
    headers: {
      "user-agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36",
      accept: "*/*",
      ...(origin ? { referer: origin } : {}),
      ...(input.headers ?? {}),
    },
  });
  if (!response.ok) {
    throw new Error(`Download fail (${response.status}): ${url.slice(0, 120)}`);
  }

  const declared = Number(response.headers.get("content-length") ?? 0);
  if (declared > MAX_DOWNLOAD_BYTES) {
    throw new Error(`File bahu moti che (${Math.round(declared / 1024 / 1024)}MB)`);
  }

  const data = Buffer.from(await response.arrayBuffer());
  if (data.length > MAX_DOWNLOAD_BYTES) {
    throw new Error("File bahu moti che (200MB thi vadhare)");
  }

  const mimeType =
    input.mimeType ??
    (response.headers.get("content-type") ?? "application/octet-stream")
      .split(";")[0]
      .trim();

  const { headers: _headers, ...rest } = input;
  void _headers;

  return saveMedia({
    ...rest,
    data,
    mimeType,
    filename:
      input.filename ?? (path.basename(new URL(url).pathname) || "download"),
  });
}

/** Disk + DB banne mathi kadho. */
export async function deleteMedia(asset: MediaAssetDocument): Promise<void> {
  if (asset.localPath) {
    await unlink(asset.localPath).catch(() => undefined);
  }
  await MediaAsset.deleteOne({ _id: asset._id });
}

/** Disk par ni file ketli moti che. */
export async function mediaSize(asset: MediaAssetDocument): Promise<number> {
  if (asset.bytes) return asset.bytes;
  if (asset.localPath && existsSync(asset.localPath)) {
    return (await stat(asset.localPath)).size;
  }
  return 0;
}

File: python/automarketing/media/store.py
"""
Media store — badhi image / video / audio file ahiya thi j pasar thay che.

Dareak file be jagya e rahe che:
  disk   — `storage/media/...` — render pipeline ne local file joiye che
  public — Catbox/Cloudinary/tunnel — Meta ne download karva mate URL joiye

Etle "save karo" ek j call che, ane public URL jarur pade tyare j bane che.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

from bson import ObjectId

from ..config import media_dir, settings
from ..db import media as media_collection
from ..db import connect
from ..errors import UserError
from ..pipeline.http import request_bytes
from . import images as imagelib
from .hosts import UploadInput, upload_public

# ------------------------------------------------------------------ #
#  Helpers
# ------------------------------------------------------------------ #

EXTENSIONS = {
    "image/jpeg": ".jpg",
    "image/jpg": ".jpg",
    "image/png": ".png",
    "image/webp": ".webp",
    "image/gif": ".gif",
    "video/mp4": ".mp4",
    "video/quicktime": ".mov",
    "video/webm": ".webm",
    "audio/mpeg": ".mp3",
    "audio/mp4": ".m4a",
    "audio/wav": ".wav",
    "audio/ogg": ".ogg",
}

MAX_DOWNLOAD_BYTES = 200 * 1024 * 1024  # 200MB


def extension_for(mime_type: str, fallback: str = ".bin") -> str:
    return EXTENSIONS.get(mime_type.split(";")[0].strip().lower(), fallback)


def kind_for(mime_type: str) -> str:
    mime = mime_type.lower()
    if mime.startswith("image/"):
        return "image"
    if mime.startswith("video/"):
        return "video"
    if mime.startswith("audio/"):
        return "audio"
    return "other"


def safe_name(name: str) -> str:
    """Path traversal band — fakt basename ane surakshit akshar."""
    base = Path(name).name
    cleaned = "".join(ch if (ch.isalnum() or ch in "._-") else "_" for ch in base)
    return cleaned[:80] or "file"


def now() -> datetime:
    return datetime.now(timezone.utc)


# ------------------------------------------------------------------ #
#  Save
# ------------------------------------------------------------------ #


async def save_media(
    data: bytes,
    *,
    mime_type: str,
    filename: str = "",
    role: str = "other",
    brand_id: Optional[str] = None,
    provider: str = "",
    prompt: str = "",
    duration: Optional[float] = None,
    width: Optional[int] = None,
    height: Optional[int] = None,
    make_public: bool = False,
) -> dict:
    """Bytes ne disk par lakhe, DB ma record banave, ane asset pacho aape."""
    await connect()

    kind = kind_for(mime_type)
    ext = extension_for(mime_type, Path(filename).suffix or ".bin")
    asset_id = ObjectId()

    base = safe_name(filename or f"{role}{ext}")
    if not base.lower().endswith(ext.lower()):
        base = f"{base}{ext}"
    stored = f"{uuid.uuid4().hex[:12]}-{base}"

    bucket = now().strftime("%Y-%m")
    directory = media_dir() / bucket
    directory.mkdir(parents=True, exist_ok=True)

    local_path = directory / stored
    local_path.write_bytes(data)

    if kind == "image" and (not width or not height):
        meta = imagelib.info(data)
        if meta:
            width, height = meta.width, meta.height

    document = {
        "_id": asset_id,
        "brand_id": brand_id,
        "kind": kind,
        "role": role,
        "filename": stored,
        "mime_type": mime_type,
        "bytes": len(data),
        "local_path": str(local_path),
        "public_url": None,
        "host": "local",
        "width": width,
        "height": height,
        "duration": duration,
        "provider": provider,
        "prompt": prompt[:2000],
        "expires_at": None,
        "created_at": now(),
    }

    await media_collection().insert_one(document)

    if make_public:
        await ensure_public_url(document)

    return document


async def save_from_url(
    url: str,
    *,
    filename: str = "",
    mime_type: str = "",
    role: str = "other",
    brand_id: Optional[str] = None,
    provider: str = "",
    prompt: str = "",
    duration: Optional[float] = None,
) -> dict:
    """Bahar na URL par thi file lai ne store ma mukhe."""
    if not url.lower().startswith(("http://", "https://")):
        raise UserError(f"Khotu URL: {url[:120]}")

    # `request_bytes` Referer ane browser User-Agent aapoaap mokle che —
    # ghani site (dakhla: ccMixter) vagar 403 aape che.
    data = await request_bytes(url, timeout=300.0)

    if len(data) > MAX_DOWNLOAD_BYTES:
        raise UserError(f"File bahu moti che ({len(data) // 1024 // 1024}MB)")

    if not mime_type:
        # Extension par thi andaj — server no content-type bharoso layak nathi.
        suffix = Path(url.split("?")[0]).suffix.lower()
        reverse = {v: k for k, v in EXTENSIONS.items()}
        mime_type = reverse.get(suffix, "application/octet-stream")

    return await save_media(
        data,
        mime_type=mime_type,
        filename=filename or safe_name(Path(url.split("?")[0]).name or "download"),
        role=role,
        brand_id=brand_id,
        provider=provider,
        prompt=prompt or url,
        duration=duration,
    )


# ------------------------------------------------------------------ #
#  Read
# ------------------------------------------------------------------ #


async def get_asset(asset_id: Any) -> Optional[dict]:
    await connect()
    try:
        oid = ObjectId(str(asset_id))
    except Exception:  # noqa: BLE001
        return None
    return await media_collection().find_one({"_id": oid})


async def get_assets(asset_ids: list[Any]) -> list[dict]:
    """Ghani asset ek saathe — ane USER e aapel KRAM ma pachi aape."""
    await connect()

    oids = []
    for value in asset_ids:
        try:
            oids.append(ObjectId(str(value)))
        except Exception:  # noqa: BLE001
            continue

    if not oids:
        return []

    found = {doc["_id"]: doc async for doc in media_collection().find({"_id": {"$in": oids}})}
    return [found[oid] for oid in oids if oid in found]


async def read_bytes(asset: dict) -> bytes:
    """Asset ni file vanche — disk par thi, nahi to public URL par thi."""
    local = asset.get("local_path")
    if local and Path(local).exists():
        return Path(local).read_bytes()

    public = asset.get("public_url")
    if public:
        # Local file gum thai gai pan public URL che — tya thi pachi laviye.
        return await request_bytes(public, timeout=300.0)

    raise UserError(f"Media file madi nahi: {asset.get('filename')}")


async def ensure_local_path(asset: dict) -> Path:
    """Local path aapo — na hoy to public URL par thi download karo."""
    local = asset.get("local_path")
    if local and Path(local).exists():
        return Path(local)

    data = await read_bytes(asset)
    directory = media_dir() / "restored"
    directory.mkdir(parents=True, exist_ok=True)
    target = directory / asset["filename"]
    target.write_bytes(data)

    asset["local_path"] = str(target)
    await media_collection().update_one(
        {"_id": asset["_id"]}, {"$set": {"local_path": str(target)}}
    )
    return target


# ------------------------------------------------------------------ #
#  Public URL
# ------------------------------------------------------------------ #


def _still_valid(asset: dict) -> bool:
    if not asset.get("public_url"):
        return False
    expires = asset.get("expires_at")
    if not expires:
        return True
    if expires.tzinfo is None:
        expires = expires.replace(tzinfo=timezone.utc)
    # Ochha ma ochhi 2 minute baaki hovi joiye — Meta ne download karva
    # jetlo vakhat male.
    return (expires - now()).total_seconds() > 120


async def ensure_public_url(asset: dict, *, prefer: Optional[str] = None) -> str:
    """
    Asset ne public URL aapo — nahi hoy to atyare j banavo.

    Kram:
      1. Pehle thi URL che ane hju valid che → e j
      2. PUBLIC_MEDIA_BASE_URL set che → aapno potano /api/media route
         (sauthi saru — koi upload j nahi)
      3. Nahi to host chain (Catbox → Cloudinary → tmpfiles)
    """
    if _still_valid(asset):
        return asset["public_url"]

    base = settings.public_media_base_url
    if base.lower().startswith("https://"):
        url = f"{base}/api/media/{asset['_id']}"
        await media_collection().update_one(
            {"_id": asset["_id"]},
            {"$set": {"public_url": url, "host": "base-url", "expires_at": None}},
        )
        asset.update({"public_url": url, "host": "base-url", "expires_at": None})
        return url

    data = await read_bytes(asset)
    result = await upload_public(
        UploadInput(
            data=data,
            filename=asset["filename"],
            mime_type=asset["mime_type"],
            kind=asset.get("kind", "image"),
        ),
        prefer=prefer,
    )

    uploaded = result.data
    await media_collection().update_one(
        {"_id": asset["_id"]},
        {
            "$set": {
                "public_url": uploaded.url,
                "host": uploaded.host,
                "expires_at": uploaded.expires_at,
            }
        },
    )
    asset.update(
        {
            "public_url": uploaded.url,
            "host": uploaded.host,
            "expires_at": uploaded.expires_at,
        }
    )
    return uploaded.url


async def delete_media(asset: dict) -> None:
    local = asset.get("local_path")
    if local:
        try:
            Path(local).unlink(missing_ok=True)
        except OSError:
            pass
    await media_collection().delete_one({"_id": asset["_id"]})


def public_view(asset: dict) -> dict:
    """API ma bahar aapva layak saaf object."""
    return {
        "id": str(asset["_id"]),
        "kind": asset.get("kind"),
        "role": asset.get("role"),
        "filename": asset.get("filename"),
        "mime_type": asset.get("mime_type"),
        "bytes": asset.get("bytes"),
        "width": asset.get("width"),
        "height": asset.get("height"),
        "duration": asset.get("duration"),
        "public_url": asset.get("public_url"),
        "preview_url": f"/api/media/{asset['_id']}",
        "created_at": asset.get("created_at"),
    }
